### PR TITLE
[ci/github] Track external code changes (blogs, tutorials)

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This adds a GitHub workflow to track changes to code we're using in external sources (e.g. blog posts or other repositories).

If a change to a tracked file is detected in a PR, a comment is added calling out the changed file and the URI of the external resource where it's being used. Also, the label `external-code-affected` is added.

In subsequent pushes to the PR, if the changes to tracked files change, the comment is updated.

This will enable us to update external sources better. It is very easy to miss subtle changes, especially when a lot of files are changed at the same time. In result some of our external blog posts or tutorials are outdated and use stale APIs. 

With a comment and label, we can proactively update external sources or filter for them after a new ray release and update in batch.

Example PR + bot interaction: https://github.com/ray-project/ray/pull/35263

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: .github/workflows/external-code-affected.yml
Comment (can-anyscale) [inline .github/workflows/external-code-affected.yml:23]: Is it possible to put this into a script with some test cases?

**Comment by amogkam** [comment]
Brilliant!

File: .github/workflows/external-code-affected.yml
Comment (can-anyscale) [inline .github/workflows/external-code-affected.yml:78]: is this php?

**Review by can-anyscale — COMMENTED** [review]
Overall, look great to me. Just wonder if we can make the github workflow more testable by moving them into a script with test cases?

**Review by maxpumperla — APPROVED** [review]
super cool!

File: .github/workflows/external-code-affected.yml
Comment (krfricke) [reply .github/workflows/external-code-affected.yml:23]: The main problem with this is that GitHub actions protects workflows, but not script files from being tampered with.

We have to run the workflow as `pull_request_target`, which means that the (ephemeral) github token will have read/write access to the github API.

If a PR from a forked repository tries to change this workflow, it will not run - it will just run the version on the master branch instead. That's a feature of GitHub's security for workflows.

However, if we just ran a script inside the repository, this protection would not apply. The user can just change the script and execute arbitrary code with read/write access to the github API.

As an alternative, we can copy the script to somewhere else and test it there - we then just have to keep the two locations in sync.

I've did quite a bit of manual testing on the script, so right now it should be fine, but I'll leave it up to you to decide if we need more tests before landing it.

File: .github/workflows/external-code-affected.yml
Comment (krfricke) [reply .github/workflows/external-code-affected.yml:78]: It's javascript!

Comment (krfricke) [reply .github/workflows/external-code-affected.yml]: Or, I guess, github-script :-D

File: .github/workflows/external-code-affected.yml
Comment (aslonnie) [reply .github/workflows/external-code-affected.yml:23]: reference link:
https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target

Comment (aslonnie) [reply .github/workflows/external-code-affected.yml:23]: > in the context of the base of the pull request

this would mean that we cannot patch vulnerability?

and the script now feels quite a bit complicated.

can we tell it to download a script/program from a source that we control and just run that?

or maybe just build our own github app..

Comment (krfricke) [reply .github/workflows/external-code-affected.yml:23]: We can develop in branches at the main Ray repo, which will also trigger the workflow on PR. The context protection only affects PRs from forked repositories.


E.g. create a branch `workflow-dev` and then file a PR for that branch. That's also what I do in this very PR btw - it's a branch on the main repository (`ci/track-external-code`) and [this PR](https://github.com/ray-project/ray/pull/35263) is a PR targeting that branch (that will be auto-rebased onto master once this PR is merged).

**Review by can-anyscale — APPROVED** [review]
Let's to this! I would love to it more if we can make it more testable, or not have to use two different workflow framework (buildkite vs. github workflow), but this protection is important so accept to unblock

**Review by richardliaw — APPROVED** [review]
(codeowner stamp)

**Pull request merged**
